keps/sig-node/4639-oci-volume-source/README.md

```diff
@@ -0,0 +1,940 @@
+   - Ensure that pulling and storing these artifacts is as efficient and secure as with OCI images.
+
+**Lifecycling and Garbage Collection:**
+   - Reuse the existing kubelet logic for managing the lifecycle of OCI objects.
+
+**Artifact-Specific Configuration:**
+   - Introduce new configuration options to handle the unique requirements of different types of OCI artifacts.
+
+**Validation:**
+   - Extend validation and security checks to cover new artifact types.
+
+**Storage Optimization in the container runtime:**
+   - Develop optimized storage solutions tailored for different artifact types,
+     potentially integrating with existing storage solutions or introducing new mechanisms.
+
+#### kubelet
+
+While the container runtime will be responsible of pulling and storing the OCI
+objects in the same way as for images, the kubelet still has to manage the full
+life-cycle of them. This means that some parts of the existing kubelet code can
+be reused, for example:
+
+- The logic how to ensure that an image exists on the node:
+  https://github.com/kubernetes/kubernetes/blob/39c6bc3/pkg/kubelet/images/image_manager.go#L102
+- The retrieval of available secrets for a pod:
+  https://github.com/kubernetes/kubernetes/blob/39c6bc3/pkg/kubelet/kubelet_pods.go#L988
+
+Specifying a `pullPolicy` be supported in the same way as for the existing
+`imagePullPolicy` API.
+
+For registry authentication purposes the same logic will be used as for the
+container image.
+
+#### CRI
+
+The CRI API is already capable of managing container images [via the `ImageService`](https://github.com/kubernetes/cri-api/blob/3a66d9d/pkg/apis/runtime/v1/api.proto#L146-L161).
+Those RPCs will be re-used for managing OCI artifacts, while the [`ImageSpec`](https://github.com/kubernetes/cri-api/blob/3a66d9d/pkg/apis/runtime/v1/api.proto#L798-L813)
+will be extended to mount the OCI object to a local path:
+
+```protobuf
+
+// ImageSpec is an internal representation of an image.
+message ImageSpec {
+    // …
+
+    // Absolute local path where the image/artifacts should be mounted to.
+    string mount_path = 20;
+}
+```
+
+This allows to re-use the existing kubelet logic for managing the OCI objects,
+with the caveat that the new `VolumeSource` won't be isolated in a dedicated
+plugin as part of the existing [volume manager](https://github.com/kubernetes/kubernetes/tree/6d0aab2/pkg/kubelet/volumemanager).
+
+#### Container Runtimes
+
+Container runtimes need to support the new `mount_path` field, otherwise the
+feature cannot be used. The kubelet will verify if the mount path actually
+exists on disk to check the feature availability, because Protobuf will strip
+the field in a backwards compatible way for older runtimes. Pods using the new
+`VolumeSource` combined with a not supported container runtime version will fail
+to run on the node.
+
+For security reasons, volume mounts should be done using the [`noexec`] flag.
```

Comment:
```suggestion
For security reasons, volume mounts should set the [`noexec`] and `ro` options by default.
```

Opening this as a separate discussion, but it would be reasonable to make these [read-only](https://github.com/kubernetes/enhancements/pull/4642/files#r1626748582) by default.

Reply:
Yep, done